Startup: Remove getopt parsing in shell script, use java CLITool

spinscale opened the pull request:
In order to ensure, we have the same experience across operating systems
and shells, this commit uses the java CLI parser instead of the shell
getopt parsing to parse arguments.

This also allows for support for paths, which contain spaces.

Also commons-cli depdency was upgraded to 1.3.1 and tests have been added.

Changes
- new exit code, OK_AND_EXIT, allowing to tell the caller to exit, as everything
  went as expected (e.g. when running a version output)

BWC breaking:
- execute() returns an ExitStatus instead of an integer, otherwise there is no
  possibility to signal by a command, if the JVM should be exited after a run.
  This affects plugins, that have command line tools
- -v used to be version, but is a verbose flag by default in the current CLI infra,
  must be -V or --version now
- -X has been removed - the current implementation was useless anyway, as
  it prefixed those properties with "es.". You should use
  ES_JAVA_OPTS/JAVA_OPTS for JVM configuration

TODO
- Can we fork into the background with java code, so we do not need that shell script hack?
- Should we rename the `-v` parameter and only support `--verbose` to resolve confusion?

MaineC commented:
Wrt. to forking into the background: Would https://commons.apache.org/proper/commons-daemon maybe be helpful here?

spinscale commented:
@MaineC looks platform specific after a quick look - would love to not have that. I need to dully understand first, what we do in in `Bootstrap()` constructor, where a non-daemonized keep-alive thread is started. Pointers appreciated :)

colings86 commented on core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java:
```diff
@@ -0,0 +1,168 @@
+                .options(
+                        optionBuilder("d", "daemonize").hasArg(false).required(false),
+                        optionBuilder("p", "pidfile").hasArg(true).required(false),
+                        optionBuilder("V", "version").hasArg(false).required(false),
```
I would say that we should have `-v` as verbose as it stands now and just support `--version` for the version parameter and not have a short form for it (if thats possible). If not, I think this is a good solution. `-v` is fairly universal as a flag for verbose.

spinscale replied:
the main thing is, that everyone reusing the CliTool infra from ES will have -v as verbose and thus I do not want to change it either... I think `-V` and `--version`

I just did a quick check...

* node/npm have -v as version
* rpm has -V
* ab has -V
* dpkg has --version

so everything is mixed. I think it is ok as long as we mention this change explicetly

colings86 replied:
agreed

colings86 commented on core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java:
```diff
@@ -0,0 +1,168 @@
+
+        private static final CliToolConfig.Cmd CMD = cmd(NAME, Start.class)
+                .options(
+                        optionBuilder("d", "daemonize").hasArg(false).required(false),
```
We have descriptions in the shell script for all these options before. Should we add them here?

spinscale replied:
Yes, thats, what the help files are for.. forces you to add proper docs :-)

colings86 replied:
So do we need the `.desc()` on the `D` option below then if we have the help files?

spinscale replied:
no, that line can be cleaned up

colings86 commented on core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java:
```diff
@@ -0,0 +1,168 @@
+            System.setProperty("es.foreground", "true");
+            if (cli.hasOption("d")) {
+                System.clearProperty("es.foreground");
+            }
```
I assume this is done because it wouldn't work to do 

```java
if (!cli.hasOption("d")) {
    System.setProperty("es.foreground", "true");
}
```

But I'm curious why?

spinscale replied:
because of a weird parsing in `Bootstrap` that I just fixed to make this more readable...

The code in bootstrap did not explicitely check for `false` or `true` but for existence to decide if elasticsearch should be run in the foreground or background. Made this implicit now.

colings86 commented on core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java:
```diff
@@ -0,0 +1,168 @@
+            if (cli.hasOption("D")) {
+                Properties properties = cli.getOptionProperties("D");
+                for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+                    System.setProperty("es." + entry.getKey().toString(), entry.getValue().toString());
```
Is it valid to specify and option as `-Des.foo.bar=baz`? If so we should probably check if the property already starts with `es.` so we don't end up with `es.es.foo.bar-baz`?

spinscale replied:
hm, how about issueing a warning on the terminal and start, rather than trying to do magical things?

colings86 replied:
sure, that would work too. I actually thought you had to specify `-Des.foo.bar` on the command line at the moment but I could be completely wrong there?

spinscale replied:
We already prepend `es.`.. see [here](https://github.com/elastic/elasticsearch/blob/master/core/bin/elasticsearch#L198)

colings86 commented on core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java:
```diff
@@ -0,0 +1,168 @@
+                        throw new IllegalArgumentException("Parameter [" + arg + "] needs value");
+                    }
+                }
+                iterator.remove();
```
very minor, but o we need to remove form the iterator here? seems like it wouldn't matter if we left the entry in the list?

colings86 replied:
same thing in the second arg check above

spinscale replied:
true, I can leave the arguments as they are and just iterate through them.

colings86 commented on core/src/main/java/org/elasticsearch/common/cli/CliToolConfig.java:
```diff
@@ -189,8 +198,17 @@ public Builder optionGroups(OptionGroupBuilder... optionGroupBuilders) {
+             * of throwing an exception
+             */
+            public Builder lenient(boolean lenient) {
+                this.stopAtNonOption = lenient;
```
Should `lenient` not be inverted. If `lenient` is `true` I would expect it to not stop at a non option so `this.stopAtNonOption` should be `false`?

spinscale replied:
Interesting, that slipped.. somewhere the meaning gets inverted, need to debug that

colings86 commented on core/src/main/resources/org/elasticsearch/bootstrap/elasticsearch-start.help:
```diff
@@ -0,0 +1,26 @@
+
+    --property=value             Configures an elasticsearch specific property, like --network.host 127.0.0.1
+    --property value
+
```
Maybe this makes my comment on descriptions above obsolete?

spinscale replied:
yes, I hope :-)

colings86 commented:
@spinscale left some comments but I think in general it looks pretty good.

spinscale commented on core/bin/elasticsearch:
```diff
@@ -132,120 +121,15 @@ case `uname` in
-
-# Start up the service
-launch_service "$pidfile" "$daemonized" "$properties"
+daemonized=`echo $* | grep -- -d`
```
this only acts on the short option... either find a workaround or also check the long one

spinscale commented:
pushed a new commit, addressing the commetns

colings86 commented:
Is there a test to check the `lenient`setting? If not, I think we should add one. Otherwise LGTM

spinscale commented:
@colings86 I actually decided to overwrite `-v` for `--verbose` and keep using `-v` for `--version` in order to be bwc compatible. You think that makes sense?

I also renamed the `lenient` option to the name of the CLI parser and fixed the documentation so it is less confusing and resembles the functionality of the commons parser

colings86 commented:
@spinscale I don't have a strong opinion on the `-v` option so happy to go with what you think is best but this will probably be our best chance of changing to be in line with other CLI tools, so now is a good opportunity if we want to change it at some point.

colings86 commented:
LGTM though

clintongormley commented:
Personally, I think I'd expect `-v` to be for verbose, as it is more common requirement than version.

spinscale commented:
ok, made `-v` for verbose again and '-V' for the version, added notes in the migration file

colings86 commented:
LGTM

spinscale committed:
Startup: Remove getopt parsing in shell script, use java CLITool

In order to ensure, we have the same experience across operating systems
and shells, this commit uses the java CLI parser instead of the shell
getopt parsing to parse arguments.

This also allows for support for paths, which contain spaces.

Also commons-cli depdency was upgraded to 1.3.1 and tests have been added.

Changes

* new exit code, OK_AND_EXIT, allowing to tell the caller to exit, as everything
  went as expected (e.g. when running a version output)

BWC breaking:

* execute() returns an ExitStatus instead of an integer, otherwise there is no
  possibility to signal by a command, if the JVM should be exited after a run.
  This affects plugins, that have command line tools
* -v used to be version, but is a verbose flag by default in the current CLI infra,
  must be -V or --version now
* -X has been removed - the current implementation was useless anyway, as
  it prefixed those properties with "es.". You should use
  ES_JAVA_OPTS/JAVA_OPTS for JVM configuration

s1monw commented:
this is so good to see! thanks @spinscale you are my hero!
